### ISSUE TITLE
Alphabetized Autocomplete

### ISSUE THREAD

**Issue by andrewtavis** [scribe-org/Scribe-iOS#188]
### Terms

- [X] I have searched [open and closed feature requests](https://github.com/scribe-org/Scribe-iOS/issues?q=is%3Aissue+label%3Afeature)
- [X] I agree to follow Scribe-iOS' [Code of Conduct](https://github.com/scribe-org/Scribe-iOS/blob/main/.github/CODE_OF_CONDUCT.md)

### Description

This issue is to add an initial auto-suggest feature to Scribe, and is the first part of #3. Work for this will be done in relation to #187 for the [Wikimania Hackathon](https://wikimania.wikimedia.org/wiki/Hackathon).

This issue comprises:
- [x] Finding resources for a function to find the next words alphabetically
- [x] Figuring out which words should be suggested to the user
    - At the start, just nouns
- [x] Removing the command bar unless the keyboard is in command mode
    - Annotations need to show up after a user presses space still
    - The coloration likely needs to change to `keyboardBgColor`, as making it clear would prevent annotations from being presented
- [x] Fixing the coloration of characters on annotations to match the new background
- [x] Properly styling the Scribe key corners/shadow when the command bar is removed
- Putting two lines between the command buttons to delineate them when they have no background (@andrewtavis)
    - [x] Add the lines to the .xib file
    - [x] Create functions that allow them to be added and removed
    - [x] Removing the left line if an annotation of long enough length is being shown
        - This needs to be done in the function to show the lines
        - Checking the length of the text in the command bar is the likely method for this
- [x] Make the command buttons user interactable when an autocomplete would be possible
    - The buttons work now, and are conditionally deactivated if an annotation is happening based on the length of the annotation
- [x] The creation of a function to find the three next words in alphabetical order to the currently typed word in the proxy
- [x] Assigning those three words to the command buttons
- [x] Running the function and doing the assignment each time a key is pressed (within `loadKeys()`)
- [x] Inserting the word on a button to the text field when a user presses an auto suggest button

### Contribution

I'll be working on this with @SaurabhJamadagni :)

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, I added something to the above, as we need to remember that even though the command bar won't be on the screen always, we're still going to need to use it to do noun and preposition annotations. Here's again the picture of eventual coloration in autocomplete:

https://user-images.githubusercontent.com/24387426/182542168-f8e337f4-4709-413c-86e4-e6ded470eed7.png

For annotation we're going to want to remove the lines, and then we can use the command bar with a transparent background to still do annotation as before (none of the command buttons would appear). Remember that the difference here is that annotation happens after space, so we won't be showing suggestions at this point yet (that's autosuggest, rather than autocomplete).

**Comment by SaurabhJamadagni** [comment]
@andrewtavis, I'm not clear on what you mean by annotations after space. Could you give me an example? Do you mean that the word suggestions will show up alphabetically even if the user just adds a space at the start?

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni :) What I mean by this is the noun genders that are shown to the user after they press space. So as the user is typing a word we want to show them autocomplete suggestions, and after they’ve pressed space we show them annotations as we do now. Rather than being “on the command bar,” they’ll be next to the Scribe key and still will be the bar text, but the command bar will be transparent.

**Comment by SaurabhJamadagni** [comment]
Ohhh I understand! So basically, we are just going to make the command bar transparent. When the user is typing, the suggestions will be shown with the vertical dividers. At other times, our prompts will be shown like we have right now. Is that correct @andrewtavis?

**Comment by SaurabhJamadagni** [comment]
Also @andrewtavis, while playing around with the command bar, I came across a bug right now. Can you check if you are experiencing it too? Basically the prompt that says what the command does, doesn't change colour when you switch appearance modes. But if you exit command mode and go back again, then it works fine. The placeholder text does change colour though. 

Steps to replicate it:
1. Select any command.
2. Change from light to dark mode or vice-versa without typing so that the prompt stays.

Images:
Before changing appearance mode:
![IMG_2957](https://user-images.githubusercontent.com/68957712/184291819-2913f1a1-6f28-4957-b3d0-37e95be89306.jpg)

After changing appearance mode:
![IMG_2958](https://user-images.githubusercontent.com/68957712/184291846-9d2209b7-a3af-44ad-bab5-d5a537796a4e.jpg)

After pressing 'x' and choosing the command again:
![IMG_2959](https://user-images.githubusercontent.com/68957712/184291937-2b12094d-18ee-4650-8734-df848762bc99.jpg)

Seems like a minor bug which we could get to later. Unless it's just a my device glitch 😅

**Comment by andrewtavis** [comment]
> Ohhh I understand! So basically, we are just going to make the command bar transparent. When the user is typing, the suggestions will be shown with the vertical dividers. At other times, our prompts will be shown like we have right now. Is that correct @andrewtavis?

Exactly :) Sorry for being a bit confusing in the explanation! Hard to wrap around all the little things to change.

Per bug that you found, I can reproduce as well. It's likely something simple like making sure that the colors are being changed within [checkDarkModeSetColors](https://github.com/scribe-org/Scribe-iOS/blob/6b9889d6a1d658a7f83b7fdd3062d203da1ba356/Keyboards/KeyboardsBase/ColorVariables.swift#L33). Dark and light mode switching was the first major bugs for Scribe - basically didn't work 🤦‍♂️🤣 Not surprised that there's an issue, and we should definitely fix it, but then it's also very unlikely that a user will experience it.

Do you want to make the issue, or should I? If you come across something, feel free to open a bug. If you're seeing it, then it should at least be checked, which is worth an issue 😊

**Comment by SaurabhJamadagni** [comment]
> Dark and light mode switching was the first major bugs for Scribe - basically didn't work 🤦‍♂️🤣 Not surprised that there's an issue, and we should definitely fix it

I see 😂 Let's fix it but definitely not a priority.

> Do you want to make the issue, or should I? If you come across something, feel free to open a bug. If you're seeing it, then it should at least be checked, which is worth an issue 😊

I'll create the issue. A little experience with that too. Never raised an issue before haha. Thanks for confirming! 😊

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, here are some directions based on what I've been working on:

I think the function to assign the buttons should be called within `conditionallySetAutoActionBtns`, which is ran after each key press. The function should find the next three nouns, and that function will assign the words to the buttons that need to be set. I'll look into making `case "AutoAction1"`, etc conditions within `executeKeyActions` so that we're ready when the function's up and running :) That should be pretty simple 😊

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, I think that the above is about as much as I should do on this to make sure that you can have a substantial contribution :) 3dc751d allows the keys to be assigned labels in `conditionallySetAutoActionBtns`, and there are now cases in `executeKeyActions` to insert the key text, add a space, and check for annotation 😊

If we can't check in before the presentation (I think that what we have so far is presentable btw), then I'm seeing in the [schedule](https://wikimania.wikimedia.org/wiki/Hackathon/Schedule) that the open hacking space is available right after. We can jump in there and talk things over, or also do a call on Google Meet :) :)

Hope you're having a nice time with your family!

**Comment by SaurabhJamadagni** [comment]
Thanks @andrewtavis! You were able to get so much done 🤯

I was playing around in Playgrounds to test out different approaches to the function and I think I have something that might work. I'll get to making that addition. Thanks for all the changes you made! I'll check back in once the changes are made. If I am not wrong, I have to checkout the feature branch you have created and push my commits to the same branch right? Or do I create a new branch? Again, sorry for these doubts right now. They are definitely not adding to the productivity 😅

**Comment by andrewtavis** [comment]
Welcome and thank you, @SaurabhJamadagni! It was nice to really code in here agin 😇

Checking out the feature branch and pushing your commits to the same one is exactly what you should do :) Not sure if you need/want to, but you can also review/accept the changes in the PR if you want to understand the changes I made. Up to you, as we can just test later and figure it out!

Looking forward to the function! I doubt that I'll add that into the presentation given timing, but if we can get it done by the showcase, then that'd be 🔥

Let me know if you have questions or need any help!

**Comment by SaurabhJamadagni** [comment]
Function to find the next three words

```
func getAutocompleteWords() {
  let keysOfNouns = Array(nouns!.keys).sorted()
  if let inString = proxy.documentContextBeforeInput {
    let stringOptions = keysOfNouns.filter { item in
         let stringMatch = item.lowercased().range(of: inString.lowercased())
        return stringMatch != nil ? true : false
    }
    
    if stringOptions.count <= 3 {
        completionWords = [String](stringOptions)
    } else {
        var i = 0
        var threeWords = [String]()
        while i < 3 {
            threeWords.append(stringOptions[i])
            i += 1
        }
        completionWords = threeWords
    }
  }
}
```

**Comment by andrewtavis** [comment]
Sorry for what likely are some confusing PR notifications, @SaurabhJamadagni :) It's all figured out now. Next time I'll know to first merge your branch into mine and then go from there.

As I mentioned, before tomorrow I'm going to try to work on the latency a bit by creating a sorted array with words that lack hyphens. I'm thinking that doing that might speed up the assignment a bit, as as of now we're creating and sorting a very large array every time we want to update the keys. It'd also be good if we always assign the "next thing in the list" even if there's nothing that comes after. Would be good to basically just fill in the next word in the list in this case.

If you have suggestions, then I'd be happy to hear them! Thanks again and again 😊

**Comment by andrewtavis** [comment]
Just played around some more 😊 This is so great! 🥳

**Comment by andrewtavis** [comment]
And the latency seems to be much faster coming out of a pre-created array and not making/sorting it each time 🙌

**Comment by SaurabhJamadagni** [comment]
> As I mentioned, before tomorrow I'm going to try to work on the latency a bit by creating a sorted array with words that lack hyphens. I'm thinking that doing that might speed up the assignment a bit, as as of now we're creating and sorting a very large array every time we want to update the keys. It'd also be good if we always assign the "next thing in the list" even if there's nothing that comes after. Would be good to basically just fill in the next word in the list in this case.

Yes @andrewtavis, I also noticed that as the bottleneck. I was wondering maybe that I could just create the array in `CommandVariables` just like we decode the nouns from the JSON there. That way we can just reference it without sorting it again and again like you said. Creating the array just once seems like the way to go. 

Thanks for the review! Yes the emails from the PR notifications were quite confusing haha. The first thought that came to mind was that I broke something irreversibly 😂

**Comment by andrewtavis** [comment]
Committing the minor change to make the array directly in 100951e :) Some current thoughts on this:

- We have an issue where words are sometimes being lower cased when they shouldn't be
- Autocomplete doesn't continue after the user has entered space, so we need to potentially split the proxy by spaces and select the second to last word to them run the word generation over (the last word when splitting by spaces will be an empty string)

We already found a very simple very very German word that wasn't in Wikidata just from me trying it out 😊 [Flohmarkt](https://www.wikidata.org/wiki/Lexeme:L691201) - German for Fleamarket or Saturday Market (they're HUGE here) - has now been added to Wikidata :)

**Comment by andrewtavis** [comment]
Specifically for the autocomplete stopping, if you just type the word and press space yourself, then it stops. If you select the word, then it tends to keep going. And in regards to the first point above, it seems that after the first autocomplete the next suggestions are all lower case, which specifically for German is problematic as they all need to be capitalized :)

**Comment by SaurabhJamadagni** [comment]
> Specifically for the autocomplete stopping, if you just type the word and press space yourself, then it stops. If you select the word, then it tends to keep going. And in regards to the first point above, it seems that after the first autocomplete the next suggestions are all lower case, which specifically for German is problematic as they all need to be capitalized :)

Ah, I didn't know that about German. Totally my bad. I just formatted it using rules used by English. That was a terrible assumption. As for the second issue, I think maybe adding the boolean variables `autoAction1Visible` and `autoAction2Visible` in the `case: spaceBar` might fix it. I'll try out these changes in the morning if that's ok @andrewtavis?

**Comment by andrewtavis** [comment]
I should have seen it in the code review myself, so no stress on the German nouns bit :) Has been a long day, so it wasn't my best review. Feel free to send along some changes in the morning! Thanks for all your hard work today! 💪

**Comment by andrewtavis** [comment]
Big thing is - now looking at Spanish where if you start typing it will start suggesting words that are lower case when you have shift pressed - it should read through on if the shift key has been pressed, and if so the words in the completions should also be capitalized :)

This is so cool 😊

**Comment by SaurabhJamadagni** [comment]
Hi @andrewtavis! Just wanted to give an update since I haven't pushed any changes since morning. I fixed the capitalisation issue and the autocomplete on space issue. After that noticed that the auto complete also went away if you delete a word afterwards as the `pastStringInTextProxy` is never updated. Working on that currently. Trying to find some way of being able to track when to change the variable so auto complete works on delete as well :)

**Comment by andrewtavis** [comment]
No stress, @SaurabhJamadagni! Just checking on the capitalization, this means that if the user has shift pressed the words will all be capitalized? Also, could we add something similarly where if capslock is on that the words are all caps? Not that I need it, but it is an expected feature :)

Keep up the good work!

**Comment by SaurabhJamadagni** [comment]
> Just checking on the capitalization, this means that if the user has shift pressed the words will all be capitalized? Also, could we add something similarly where if capslock is on that the words are all caps? Not that I need it, but it is an expected feature :)

Yes, the suggestions show a capitalized word if the shift is pressed. I could try it out with the capslock as well no problem. I'll get to it in just a bit.

**Comment by andrewtavis** [comment]
Wonderful! This is great! It'd be cool to be able to show that off, and that way we can also show autocomplete in other keyboards :)

**Comment by SaurabhJamadagni** [comment]
Hey @andrewtavis, I fixed the no auto complete on delete and added the capslock word suggestions as well. But when the text proxy is empty, repeatedly pressing the delete button is crashing the keyboard. I'll push my commits and create a PR right now. Could you maybe check why that might be happening. Maybe something you have experienced before. 😅

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni :) I'll give it a look and potentially just bring it in for us to work on later in a different bug issue that needs to be resolved before the release. Will let you know how things are looking in the PR.

**Comment by andrewtavis** [comment]
Thanks again for all this, @SaurabhJamadagni 😊 You've worked a ton on Scribe this weekend! You really have my thanks, and I hope the experience has been an interesting one for you :) I mentioned in the PR that capitalization still needs a bit of work, but it's almost exactly what we need. 

I'm thinking also that maybe adding in verbs and prepositions might be best, as that way there's always suggestions happening :) We can work on this going forward and shoot for a release with #194 when we're feeling comfortable with how it all's working. Major major strides on the project this weekend, and it's totally amazing how much we've gotten done in a short time!

**Comment by SaurabhJamadagni** [comment]
Hey @andrewtavis! How are you doing? 😊

I will get to the requested review on PR #195 sometime today or tomorrow. I'm also interested in working on the AutoSuggest issue so I'll comment on that page shortly. Looking forward to the next release! It's super exciting 🚀

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni :) I'm well! How's all with you? Am talking with someone from the hackathon about whether they'd like to work on the Android version :)

No stress on the PR for #195 - get to it when you can. And yes, let's definitely talk #194 so that we can get both of these features released! I'm really looking forward to it as well, and think that people are going to be really impressed 😊

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni, hope you're well :) 

Let me know if you need any support on the PR. I was feeling a bit tired post hackathon and had lots to do, but have a bit more bandwidth now if you want to discuss the needed changes 😊

**Comment by SaurabhJamadagni** [comment]
Hi @andrewtavis! Sorry I haven't been updating lately. My summer break ended today. So I have was a bit busy prepping for classes. Anyway, I have tomorrow off and a long free weekend ahead. I'll give it a shot tomorrow and let you know if I need help. Thanks for asking. Sorry for the delay! 😊

**Comment by andrewtavis** [comment]
No worries on a delay whatsoever, @SaurabhJamadagni! Just be in touch and we'll figure it out :)

**Comment by andrewtavis** [comment]
Also, a final thing we should consider is expanding the word list a bit. Linking in the potential of using [wordfreq](https://github.com/rspeer/wordfreq/), we could also potentially subset a few popular verbs and include their conjugations in the list, and maybe including prepositions would also make sense. Removing words that are only one or two characters might also make sense :think

**Comment by andrewtavis** [comment]
Also also, I'm wondering if my logic on annotation canceling the plural autocomplete field was misplaced. Maybe the best thing is to keep it all in the first section where the translate button would be, and then we can just do `...` if the word is long. The user knows what word they just typed, so they don't need to see the whole thing. See designs below:

![Autocomplete](https://user-images.githubusercontent.com/24387426/185496717-8b127a77-b393-468e-9a4d-1649d59e9d77.png)

**Comment by SaurabhJamadagni** [comment]
> Maybe the best thing is to keep it all in the first section where the translate button would be, and then we can just do ... if the word is long. 

Yeah @andrewtavis I was wondering the same thing during the hackathon. But because of the time constraint it completely slipped my mind to discuss it. It also seems like unnecessary computation when it has to figure out if the word is short enough to fit in one button or not. We should definitely look into this appraoch where we keep it in just the translate button.

**Comment by SaurabhJamadagni** [comment]
Also, the wordfreq library that you have linked seems super cool! We could definitely utilise it in autosuggest, rather definitely should 😅

> we could also potentially subset a few popular verbs and include their conjugations in the list, and maybe including prepositions would also make sense.

I didn't understand what you mean by this. I am guessing you are referring to the `autocompleteWords` array that we use to filter words from.

**Comment by andrewtavis** [comment]
> I didn't understand what you mean by this. I am guessing you are referring to the `autocompleteWords` array that we use to filter words from.

Yes exactly, adding words to `autocompleteWords` to make sure it's covering most of the use cases :)

**Comment by andrewtavis** [comment]
> We should definitely look into this appraoch where we keep it in just the translate button.

I'll change it when you're done with #195 :) I think that this also adds the potential to allow the user to click on the annotations and have the translate button fire for other features. Specifically what I'm thinking is adding ways to figure out what's the difference if there are two genders on a noun or two cases for a preposition - the user could press that button and then have information displayed to them to help them figure out the context.

Basic example of this is the word "Schild" in German, which is masculine or neutral, and has the meaning shield and sign respectively.

**Comment by SaurabhJamadagni** [comment]
> I think that this also adds the potential to allow the user to click on the annotations and have the translate button fire for other features. Specifically what I'm thinking is adding ways to figure out what's the difference if there are two genders on a noun or two cases for a preposition - the user could press that button and then have information displayed to them to help them figure out the context.

That is a nice idea! This could be a feature that works when the user is online though. Adding this amount of size could surely add up, increasing the size of the app I think. What do you think? 😅

**Comment by SaurabhJamadagni** [comment]
@andrewtavis, I also noticed a bug showing similar nature as when the keyboard crashes when delete is pressed. It isn't allowing to highlight text and then delete it. To replicate it, try and select a bunch of text and hit the delete on the scribe keyboard. The keyboard just bobs up and down, similar to when the keys would be reloaded. I think the problem which is causing the crash is the same problem that is causing this. It's just a hunch. Didn't create a new issue around it because it is working fine in the app. Just a problem with our current code state.

**Comment by andrewtavis** [comment]
> That is a nice idea! This could be a feature that works when the user is online though. Adding this amount of size could surely add up, increasing the size of the app I think. What do you think? 😅

Yes certainly, but then the goal for later is that the user will be downloading language packs, so this could still work offline potentially :)

**Comment by andrewtavis** [comment]
> To replicate it, try and select a bunch of text and hit the delete on the scribe keyboard.

Thanks for pointing this out, @SaurabhJamadagni! Will have a look when I'm check #195 :)

**Comment by andrewtavis** [comment]
Also @SaurabhJamadagni, when you have a moment let me know what you think of the current ideas within #194 and if you have questions :) I was thinking that you could do the Swift side implementation, and I can generate the words it'll use based on what we want to build. Obviously just random words would be easiest, but if we can create a `commonWordSuggestions` array easily enough with Python, then it could definitely be worth it to have this feature be close to optimal from the start 😊

**Comment by SaurabhJamadagni** [comment]
Sorry @andrewtavis! I didn't have my email notifications on for issue #194. So didn't have an idea you had commented there. I'll check it out soon. I was also wondering if I could tag along for some parts when you do the python side of things. It could be a nice learning opportunity through observing the process and working with different technologies. I have a good grasp of python so I might be able to help out a bit too hopefully 😅

Do let me know whenever and if it's possible :)

**Comment by andrewtavis** [comment]
100% yes, @SaurabhJamadagni :) We can figure out another time for a call and then I can walk you through the process. You’d of course also be welcome to contribute, so I’ll make some tasks like in here 😊

Big question is, do we do the simple one with a series of most common words that we assign randomly, or do we think that the other method with Wikipedia generated arrays of words that follow common words would be a better option? I don’t think it’ll be crazy amounts of effort to set up, as again I have a lot of the code in other projects 🐍  There definitely will be some waiting involved though for all the calculations 😅

**Comment by SaurabhJamadagni** [comment]
> We can figure out another time for a call and then I can walk you through the process. You’d of course also be welcome to contribute, so I’ll make some tasks like in here 😊

Thank you so much @andrewtavis! I would love that :)

> Big question is, do we do the simple one with a series of most common words that we assign randomly, or do we think that the other method with Wikipedia generated arrays of words that follow common words would be a better option? 

I commented on issue #194 with my thoughts on this just now. I think the Wikipedia system would give us better results and would be more practical. I still don't quite get how we will implement it as I said in the other comment. But this seems like a good direction to go in for now. We can go back to a frequency based implementation if we face problems of course.

**Comment by andrewtavis** [comment]
I’ll comment more in the other issue, @SaurabhJamadagni :) Looking forward to talking some Python with you!

**Comment by andrewtavis** [comment]
Thanks again for all the efforts for this, @SaurabhJamadagni! We're now done with the initial tasks, and I'll take on two final ones:

- [ ] Remove the deactivation of the conjugate button state and change long words to their start plus `...`
- [ ] Make sure that all words that are capitalized in the JSONs are capitalized even if the user is typing in lower case, so this way German nouns and proper nouns in other languages will be presented to the user as they should be

Once we're done with these we'll be good to go! 🚀

**Comment by andrewtavis** [comment]
> Also, let me know if you would like me to try to figure this one out too @andrewtavis. I would be eager to jump on a meeting to try to solve this problem, if needed. Maybe the bouncing of ideas might work like last time. Thanks for reviewing the PR 😊

I'll get back to you on if a call is necessary, @SaurabhJamadagni :) We'll have a lot to talk about regardless. Not 100% sure on my schedule for this week, but I'll be in touch via email about a call for finishing this and the two issues for autosuggest (the one here and the one in Scribe-Data).

**Comment by andrewtavis** [comment]
Also @SaurabhJamadagni, let me know if you think that eventually removing the word from annotations and centering the icons on the space where the translate button is would be better. I agree that the conjugate button being disabled doesn't make sense, and now that I'm looking at it with the `...` for long words, I'm thinking that maybe just centering them and not having the word might be best? This would require some more edits of the `.xib` file, but might look better than having words that are and are not cut off. Furthermore, I'm not sure if having things that are left justified - the annotations - next to things that are centered - the autocompletions and suggestions - looks particularly good. 

I don't think this needs to happen now, but it might be better in the long term. I think that seeing an annotation without the word that was just typed would be enough 🤔

**Comment by SaurabhJamadagni** [comment]
> let me know if you think that eventually removing the word from annotations and centering the icons on the space where the translate button is would be better. I agree that the conjugate button being disabled doesn't make sense, and now that I'm looking at it with the ... for long words, I'm thinking that maybe just centering them and not having the word might be best? 

Sorry @andrewtavis, I am not able to imagine what you mean by removing the word and centering the icons. Is it possible to create a design mock-up? Removing the word does make sense as it is clear that the annotations are for the word that was last typed by the user. I just didn't understand how you are describing the UI would look like 😅

**Comment by andrewtavis** [comment]
> See designs below:
> 
> ![Autocomplete](https://user-images.githubusercontent.com/24387426/185496717-8b127a77-b393-468e-9a4d-1649d59e9d77.png)

@SaurabhJamadagni, I don’t have access to internet on my computer, but generally what I’m referring to is like the picture above, but say that the word Zeit wasn’t shown at all, and instead the red F annotation would be centered. So all annotations would be centered over the translate button, and if there are more then one then the group would be centered as well :)

Hope that makes sense! 😊

**Comment by SaurabhJamadagni** [comment]
Oh! Yes got it @andrewtavis. That makes sense as the only reason people would tap on the annotations would be when we later implement the feature that shows context when a word has more than one annotation. I got it. Thanks for the explanation. I am sorry if I bothered you at an inconvenient time 😅

**Comment by andrewtavis** [comment]
Not a problem, @SaurabhJamadagni! If I’m checking here it means I have the time :) Was just at the airport checking things on my phone 😊

Let’s finish this and I’ll make a new issue for switching the annotation styles. Would likely be best if we move away from the ‘.xib’ based annotations and make the views with code, as I think it’ll be difficult to position them correctly unless we do it programmatically.

This is all gonna look so great afterwards! 😊

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni :) I'm getting to all of this in this issue at the moment. Let me know when you're available for the checkin 😊

**Comment by SaurabhJamadagni** [comment]
Hey @andrewtavis! Tomorrow being a Sunday, I can join in anytime before 15:00 UTC. Or if you are available sometime before 11:00 UTC today itself? 😅

**Comment by andrewtavis** [comment]
Let’s shoot for 13 UTC tomorrow :) I’ll send along a link 😊

**Comment by SaurabhJamadagni** [comment]
Ok that works! See you then :)

**Comment by andrewtavis** [comment]
> Also @andrewtavis, while playing around with the command bar, I came across a bug right now. Can you check if you are experiencing it too? Basically the prompt that says what the command does, doesn't change colour when you switch appearance modes. But if you exit command mode and go back again, then it works fine. The placeholder text does change colour though.

@SaurabhJamadagni, just FYI, [f0bd0a0](https://github.com/scribe-org/Scribe-iOS/commit/f0bd0a0f98e48f7b7d93125fafd97feb9f7bf98b) fixed the command bar coloration issues 😊

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, an idea that I'd like your feedback on :)

I'm thinking that a lot of the issues that we have with the keyboard working are coming from the overlapping booleans that are super hard to keep track of. To fix this, I'm working on switching `commandState` into a `var` rather than a `bool`, with it being an instance of an enum `CommandState`:

```swift
/// States of the keyboard corresponding to which commands the user is executing.
enum CommandState {
  case idle // normal view of the keyboard
  case select // select a command, formerly scribeKeyState == true
  case translate // combines getTranslation, commandState == true and switchKeyboardInput
  case conjugate // combines getConjugation and commandState == true
  case selectConjugation // formerly conjugateView
  case plural // combines getPlural and commandState == true
  case alreadyPlural // formerly isAlreadyPluralState
  case invalid // formerly invalidState
}
```

The general idea is that this will be a catch all state variable that will determine what the user sees. We'd thus have much more direct control over what's being shown by simply activating the corresponding `commandState`.

I've changed most of it over in the last hour and surprisingly it already kind of works again 😅 Will need a bit more work, but I think that I can get this done over the weekend, and then we'll have much smoother sailing going forward 😊

**Comment by andrewtavis** [comment]
When this is done, I'd send along a PR that makes these changes and fixes another issue that invalid entries are now conflicting with the auto actions. We need to make it work so that the "Not in Wikidata" message is still displayed to the user after an invalid command entry, and then that and the `commandBar` disappear after the next key press to allow for autocomplete again.

After that we can work on the issues with deletion :)

**Comment by andrewtavis** [comment]
Already such a crazy improvement:

Rather than the following to reset the keyboard command state:
```swift
switchInput = false
scribeKeyState = false
commandState = false
conjugateView = false
```

We have:
```swift
commandState = .idle
```

😊

**Comment by SaurabhJamadagni** [comment]
Hi @andrewtavis! That's a great improvement. I was almost about to go on a boolean hunt as to what might be causing the crash but this seems like a great idea and would make it much simpler! Looking forward to the commit. Excited to see the ease in readability these changes will bring :)

**Comment by SaurabhJamadagni** [comment]
> You could check a bit into autocomplete not working after commands now if you'd like :)

Yes, I will try to figure this one out over the weekend as well. Also, simultaneously working on the mock version of getting the auto-suggest words from the 500 words list that we will be creating. Playing around with it a little bit in Swift Playgrounds.

I also wanted to explore debugging over this weekend a little bit. I will search up a bit on how to debug better in Xcode. That should help us track down problems much quicker hopefully 😅

**Comment by andrewtavis** [comment]
> I also wanted to explore debugging over this weekend a little bit. I will search up a bit on how to debug better in Xcode. That should help us track down problems much quicker hopefully 😅

Yes, @SaurabhJamadagni, that would be super helpful! Sorry that I'm not as much of support for debugging, but I'd be happy to learn from you on this as well. Has just been something that I didn't pick up as well as I should :)

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, a5ea9ea is a lot of refactoring that was long overdue, and does fix some issues with invalid messages not being presented properly, but we still have the unfinished points I [mentioned here](https://github.com/scribe-org/Scribe-iOS/issues/188#issuecomment-1222253453) before we can close this. A lot closer though with how much easier checking flows should be!

**Comment by SaurabhJamadagni** [comment]
> A lot closer though with how much easier checking flows should be!

Couldn't agree more @andrewtavis! Sorry for the slow responses. These are a lot of changes combining all the three issues we are currently working on. I was just taking the time to process them and make a game plan for myself. 

On a slightly unrelated note, can you explain me what the issue #201 that you have created would do? I tried to go through the documentation on UILexicon but I did not understand that clearly. Is it basically like the user creates a key 'tbh' which automatically gets expanded to 'to be honest'?

**Comment by andrewtavis** [comment]
> These are a lot of changes combining all the three issues we are currently working on. I was just taking the time to process them and make a game plan for myself.

Yes, I agree that it's kind of a lot right now, and I'm sorry for that. I'm very much looking forward to when this is all done and we can look at the issues and be like, "Which small issue that's not really related to the others would we like to work on now?" 😄😊

**Comment by andrewtavis** [comment]
I think that getting this one done will be a big weight off, @SaurabhJamadagni. As soon as we're done here, #194 should be straight forward. You again can start working on that if you'd like :)

**Comment by andrewtavis** [comment]
We're now down to these last two issues, @SaurabhJamadagni! 🥳

- [ ] Fix a bug where delete at the start of the proxy now crashes the keyboard
- [ ] Fix a likely related bug where once a user starts typing in a command delete will end the command/crash the keyboard

As I said in the above PR, if you can get the bugs for delete, that'd be really great 😊 Let me know on all this :)

**Comment by SaurabhJamadagni** [comment]
Yes @andrewtavis, I am currently working on the above. I'll try to share a small code snippet by evening that I wrote for #194 that selects the next three words from the common words array which will be a part of the JSON. I am currently reading a few articles on debugging to help me get tracking the above two bugs. I'll have an update soon hopefully 😅

**Comment by andrewtavis** [comment]
Sounds great, @SaurabhJamadagni! Feel free to share what you learn when we do the next call 😊

And just FYI, the structure of the autocomplete array will be the following:

{
"word": ["option", "option", "option"],
"word": ["option", "option", "option"],
...
}

**Comment by SaurabhJamadagni** [comment]
Hey @andrewtavis, two more minor bugs that I noticed that could be added to the list are:
1. Long words stop getting the `...` for some words.

<img width="351" alt="image" src="https://user-images.githubusercontent.com/68957712/189129640-a598ce60-ff78-4722-90dd-06e5bfdd5583.png">

2. We can't select a bunch of text and delete it all at once. It leads to the bobbing of the keyboard just like with the delete crash.

<img width="347" alt="image" src="https://user-images.githubusercontent.com/68957712/189130018-03b8eaad-d204-47dc-8ea6-71fc1c946d75.png">

I think I will need your help with the delete crash issue. I read up on debugging and some of the things are very clever but when I run the project, the code just won't stop at the breakpoints. I think it's not letting me run in debug executable mode. Trying to figure it all out. Sorry for taking so much time to fix this bug. Hoping to get it soon.

**Comment by andrewtavis** [comment]
Hi @SaurabhJamadagni, no stress on timing! I wouldn't have had time to review this week anyway as I've been very busy for work. To the above:

1. I noticed this as well, but it's going to be fixed in #197, as the words won't be showing anymore at all :)
2. Noted, and thanks for finding it! I'll add it to the list above so we remember to check for it.

I also tried a bit to figure out debugging, and it also wasn't stopping at the breakpoints... Take your time on all this - we'll get it soon! 😊 

My plan is to work a bit on #197 in a branch tomorrow. I'll start by removing the current UI elements in the `.xib`, and then I'll rewrite the functions to generate the annotation icons without the words. I'll be in the US visiting friends and family from the 16th to the 25th, but I should still have time to work on all this. My personal goal is to have v2.0.0 released by around the 25th, so two weeks :)

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, on second thought, the second point below is a feature that was never implemented :) Good work on finding it! I'll make a separate issue for it :)

**Comment by andrewtavis** [comment]
@SaurabhJamadagni, hope it's ok that I went ahead and made these fixes :) I was checking the delete codes to maybe fix the issue with selected text deletion, and then realized the issues [d69bad8](https://github.com/scribe-org/Scribe-iOS/commit/d69bad8e65ae051021c48c1db359c31f1d462127) wants to address. Figured since I'm looking at it, just see if I can get it done 😊

Hopefully merging #206 will complete this, and then we can shift to the annotation fixes and autosuggest 🚀

**Comment by SaurabhJamadagni** [comment]
>  hope it's ok that I went ahead and made these fixes :)

Absolutely @andrewtavis! I'll review the PR in a bit. I definitely want to know what was the problem as this bug really got under my skin haha! Thanks 😊

**Comment by SaurabhJamadagni** [comment]
> @SaurabhJamadagni, on second thought, the second point below is a feature that was never implemented :) Good work on finding it! I'll make a separate issue for it :)

Thanks! I think this works in the production version though that is currently in the app store. I think it stopped working during all the changes that we created. Can you check if that's the case? 😅

**Comment by andrewtavis** [comment]
> Thanks! I think this works in the production version though that is currently in the app store. I think it stopped working during all the changes that we created. Can you check if that's the case? 😅

Yes, you're right, @SaurabhJamadagni :) It works on my phone on the current version. I just checked and d69bad8 removes that bug as well. I guess I just figured that we'd need to implement something to have the selected text be deleted, but there's something going on under the hood to replace it :) I'll delete #205 😊

**Comment by andrewtavis** [comment]
Well, @SaurabhJamadagni, that was quite the issue 😊 Glad to have it done, and our path to v2.0.0 is a lot more clear :) You're welcome to start working more on functions needed for #194 (I'll write a bit in there now), and I'll start on #197 :)

**Comment by SaurabhJamadagni** [comment]
> Glad to have it done, and our path to v2.0.0 is a lot more clear :) You're welcome to start working more on functions needed for https://github.com/scribe-org/Scribe-iOS/issues/194 (I'll write a bit in there now), and I'll start on https://github.com/scribe-org/Scribe-iOS/issues/197 :) 

Awesome @andrewtavis! I am super excited about finishing this issue as well! On to the next 🚀
